sigma-ser/src/scorex_serialize.rs

```diff
@@ -0,0 +1,200 @@
+pub type ScoreXSerializeResult = Result<(), ScoreXSerializationError>;
+
+///TODO
+/// ScoreX Serializable Trait.
+pub trait ScoreXSerializable: Sized {
+    // TODO add error types
+    /// Write `self` to the `writer`
+    fn scorex_serialize<W: WriteSigmaVlqExt>(&self, w: &mut W) -> ScoreXSerializeResult;
+    /// parse `self` from `reader`
+    fn scorex_parse<R: ReadSigmaVlqExt>(r: &mut R) -> Result<Self, ScoreXParsingError>;
+
+    /// Serialize a ScoreXSerializable value into bytes
+    fn scorex_serialize_bytes(&self) -> Result<Vec<u8>, ScoreXParsingError> {
+        let mut w = vec![];
+        self.scorex_serialize(&mut w)?;
+        Ok(w)
+    }
+    /// Parse `self` from the bytes
+    fn scorex_parse_bytes(mut bytes: &[u8]) -> Result<Self, ScoreXParsingError> {
+        Self::scorex_parse(&mut bytes)
+    }
+}
+
+impl<T: ScoreXSerializable> ScoreXSerializable for Vec<T> {
+    fn scorex_serialize<W: WriteSigmaVlqExt>(&self, w: &mut W) -> ScoreXSerializeResult {
+        w.put_u32(self.len() as u32)?;
+        self.iter().try_for_each(|i| i.scorex_serialize(w))
+    }
+
+    fn scorex_parse<R: ReadSigmaVlqExt>(r: &mut R) -> Result<Self, ScoreXParsingError> {
+        let items_count = r.get_u32()?;
+        let mut items = Vec::with_capacity(items_count as usize);
+        for _ in 0..items_count {
+            items.push(T::scorex_parse(r)?);
+        }
+        Ok(items)
+    }
+}
+
+impl<T: ScoreXSerializable, const L: usize, const U: usize> ScoreXSerializable
+    for BoundedVec<T, L, U>
+{
+    fn scorex_serialize<W: WriteSigmaVlqExt>(&self, w: &mut W) -> ScoreXSerializeResult {
+        self.as_vec().scorex_serialize(w)
+    }
+
+    fn scorex_parse<R: ReadSigmaVlqExt>(r: &mut R) -> Result<Self, ScoreXParsingError> {
+        Ok(Vec::<T>::scorex_parse(r)?.try_into()?)
+    }
+}
+
+/// Corresponds to `VLQ(UInt)` format from `ErgoTree` spec.
+impl ScoreXSerializable for u32 {
+    fn scorex_serialize<W: WriteSigmaVlqExt>(&self, w: &mut W) -> ScoreXSerializeResult {
+        w.put_u32(*self)?;
+        Ok(())
+    }
+    fn scorex_parse<R: ReadSigmaVlqExt>(r: &mut R) -> Result<Self, ScoreXParsingError> {
+        let v = r.get_u32()?;
+        Ok(v)
+    }
+}
+
+impl<T: ScoreXSerializable> ScoreXSerializable for Option<Box<T>> {
+    fn scorex_serialize<W: WriteSigmaVlqExt>(&self, w: &mut W) -> ScoreXSerializeResult {
+        match self {
+            Some(v) => {
+                w.put_u8(1)?;
+                v.scorex_serialize(w)
+            }
+            None => Ok(w.put_u8(0)?),
+        }
+    }
+
+    fn scorex_parse<R: ReadSigmaVlqExt>(r: &mut R) -> Result<Self, ScoreXParsingError> {
+        let tag = r.get_u8()?;
+        Ok(if tag != 0 {
+            Some(T::scorex_parse(r)?.into())
+        } else {
+            None
+        })
+    }
+}
+
+/// serialization roundtrip
+#[allow(clippy::expect_used)]
+pub fn scorex_serialize_roundtrip<T: ScoreXSerializable>(v: &T) -> T {
```

Comment:
After doing some ripgrepping, I've realized that sigma_serialize_roundtrip is also only used in tests as well. Should I add a #[cfg(test)] attribute to it so it only gets compiled in when running tests? Or could it be useful for external users of the library as well?

Reply:
Oh, I forgot that it would be used in user's tests. So let's leave it like this.

Reply:
Hmm, should I make it user-facing again. I changed it in https://github.com/ergoplatform/sigma-rust/pull/485/commits/c78bb796f41ebcc9ebcf197ea8dc4d94e36d0e19 especially since the scorex_serialize module is not visible itself, so scorex_serialize_roundtrip would have to be imported into lib.rs as well

Reply:
Yes, please. Users should be able to call it.